compiler/rustc_error_codes/src/error_codes/E0729.md

```diff
@@ -1,30 +0,0 @@
-Support for Non-Lexical Lifetimes (NLL) has been included in the Rust compiler
```

Comment:
Please revert this change. Error code explanations should never be removed. Please only update it to mention that it's not emitted by the compiler anymore.

Reply:
Done. I don't really like this, what use are these old docs (especially considering lots aren't documented)? If your using an old compiler, you can `rustc --explain Exxxx` and if you are using an up-to-date compiler then there is literally no reason you could ever need those docs.

Reply:
Because people might still use old rustc version and look for their explanation online. If we remove them, they disappear in latest docs version, forcing you to look into the older ones (which aren't very easy to find if you don't know where to look).

Reply:
Do people ever use `--explain` or is that just me lol? It's very interesting, error codes aren't gated by edition. *Personally*, I would prefer purging all the old ones and adding a note like, "If an error code is not listed, then it has been removed in a recent `rustc` version, please update. You can find old error code documentation \[here](link) or by using the `--explain` flag." I thought that you're on your own if you don't update because of the backwards-compatibility promises.

Reply:
I meant online. Here to be exact: https://doc.rust-lang.org/error_codes/error-index.html

Reply:
Yeah, I know. But we don't have to support them? For example, they don't get bug fixes. Why should they get docs?

Reply:
Error codes aren't *bugs*, they're errors in user's code.

Reply:
What I'm saying (rather badly) is that we (all the rustc developers) do not support versions of rustc that are not up to date. If borrow checker bug is found in rustc for an old rustc version (say 1.60) then *it isn't fixed*. The fix for people who use affected versions of rustc *is to update*, this works because Rust is always backwards-compatible (there is no effort to use a newer rustc). However the "problem" I have is that this isn't true for long-form error code docs, we keep the old docs even though *users should update*. Another example: in stdlib, we don't keep docs for removed functions (however rare that is).

Reply:
We never removed a function in stdlib because it would break semver.

Also I think we're not talking about the same thing at all. Just in case: imagine a company using only rust 1.10 for whatever reason. They have a very old code and therefore see completely legit old error codes thrown out by the compiler. They can look it up locally of course, but let's say they want to ask for extra explanations about the explanation and therefore share its online location. However, since the error code is long time since then, it was removed. Then there is no way to look for help for the user.

It's somewhat similar to why we don't re-use error codes: because we want to confuse users using old error codes which would have invalid explanations.

So, the error code and its explanations are still perfectly valid for the context in which they were written. If someone is looking for its explanation, it means it has a use for it. And once again, an error code doesn't mean something is wrong with the compiler, it means it's a legit error from the user's code.

Final point: it's current in the industry to pinpoint a specific compiler version and use it for a few years before migrating to a new one. For such users, the documentation needs to remain available.

Reply:
> We never removed a function in stdlib because it would break semver.

What about editions?

Fair points, I don't entirely agree but it makes sense. 

As an aside, would you ever write documentation for error codes that are no longer emitted? Also, is this resolved then?

Reply:
> What about editions?

That cannot be done across editions. It is a very limited mechanism.

Either way, the policy is that old error codes are not removed from the index, and this isn't the right place to discuss changing the policy 😛 Consider opening a zulip thread or an issue if you feel strongly about it - but I think you'll find many share Guillaume's stance.

Reply:
> What about editions?

We never removed std libs API in editions. Deprecated for sure but removed, never.

> As an aside, would you ever write documentation for error codes that are no longer emitted?

The only goal here is to prevent to remove knowledge. Adding content for outdated errors is another matter. So the answer is no, I would not write documented for removed error codes.

> Also, is this resolved then?

You un-removed the file so all good for me.